### ISSUE TITLE
Model Plugin - cloud-instance-metadata

### ISSUE THREAD

**Issue by jawache** [Green-Software-Foundation/ief#132]
- **What**:  A model which given some input params regarding cloud instances returns information useful to downstream models for computation
- **Why**: So every model doesn't need to replicate cloud instance lookups
- **DoD**:
- [ ] Does the model meet the specification?
- [ ] Does the model have complete test coverage, including edge cases, happy path and error cases?
- [ ] Has documentation been written for someone writing an IMPL using this model?
- [ ] Has the `units.yml` been updated to take into consideration the params used in this model

## Config
- `variable`: If this variable is present in the top level config what behaviour does it trigger? What is the default value if none is provided?

## Observation
_List the variables this model expects in an input observation_
- `cloud-vendor`: Unique id of the cloud vendor, e.g. "aws"
- `cloud-region`: Region as per that cloud vendors definition, e.g. "us-west"
- `cloud-instance-type`: Id of that cloud instance for that vendor, e.g. "t1.micro"

## Impacts
_List the variables this model exports out as an impact_
- `physical-processor`: The name of the underlying physical processor, family, generation.
- `vcpus-allocated`: The number of vCPU's allocated to this instance.
- `vcpus-total`: The number of vCPUs available to the underlying physical server.

## Behavior
What does the model do to turn the observations into impacts? List any important equations or references.
- For every cloud provider given an instance type tries to determine some common information for use by other models.
- We want this data source to be incredibly up to date so it need to read from the latest data, e.g. APIs from the cloud providers or some regularly export CSV.
- This will first require a lot of investigation and analysis for where and how to get the data. This might be one source https://github.com/vantage-sh/ec2instances.info/blob/master/ec2.py but we need to investigate others

## Assumptions
What are the core assumptions this model makes?
- That we have information for every instance type.
- That all we care about right now is CPU's (we need to think about Mem and GPUs as well)

## Limitations
In what situations is this model limited in it's function, behaviour, accuracy etc...
- It needs to be upto date, **don't** use the 2 year old teads model data. All modern instance types need to be supported and instance types change.

## Example
Provide at least one example of some input config and observations. This is in pimpl format not rimpl format, so the config is the global config and all parameters must be present on the observation.

### Input

```yaml
observations: 
  - timestamp: 2023-07-06T00:00
    duration: 3600
    cloud-vendor: aws
    cloud-region: us-west
    cloud-instance-type: t3.micro
```

### Output

```yaml
observations: 
  - timestamp: 2023-07-06T00:00
    duration: 3600
    cloud-vendor: aws
    cloud-region: us-west
    cloud-instance-type: t3.micro
impacts: 
  - timestamp: 2023-07-06T00:00
    duration: 3600
    cloud-vendor: aws
    cloud-region: us-west
    cloud-instance-type: t3.micro
    physical-processor: Intel Skylake E5 2686 v5
    vcpus-allocated: 2
    vcpus-total: 24
```

**Comment by jawache** [comment]
@gnanakeethan and @jmcook1186 I'm not really back to work yet but I wanted to share this idea for a model. If @gnanakeethan is looking for some work this might be a good next one to implement. But it's low priority, focus on the other ones required for the demo(s)

**Comment by gnanakeethan** [comment]
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstanceTypes.html

**Comment by jawache** [comment]
@gnanakeethan As discussed, use the data from the CCF data base for AWS as an Alpha so we can start demoing the functionality. This will be outdated and only contain a subset of the AWS instance types but has a mapping of instance types to processor names which is what is needed.

The full implementation of this is going to be quite significant and require a human to fill in the missing data as well as a process so as new instance types are released we know to update this model.

**Comment by jawache** [comment]
Note: @gnanakeethan make the data for this match tdp if possible

**Comment by gnanakeethan** [comment]
@jmcook1186 @jawache the TDP Finder works alongside the output from the CCF data we already have.

**Comment by jawache** [comment]
@gnanakeethan - for the universe of AWS instance type mentioned here, how many of them match processor from the TDP model plugin.

**Comment by gnanakeethan** [comment]
CPU Types Listed in AWS Instances data from CCF. 

- Graviton
- Xeon E5-2651 v2
- Xeon E5-2670
- Xeon E5-2680 v2
- Xeon E5-2666 v3
- Xeon Platinum 8124M
- Xeon Platinum 8275CL
- EPYC 7R32
- Graviton2
- Xeon E5-2676 v3
- Xeon Platinum 8259CL
- Xeon E5-2670 v2
- Xeon E5-2686 v4
- Xeon Platinum 8175M
- Xeon E5-2650
- Xeon E5-2665
- EPYC 7571
- Xeon Platinum 8252C
Xeon Platinum 8375C
Core i7-8700B
Xeon Platinum 8176M
Xeon E7-8880 v3
Xeon Platinum 8151

**Comment by gnanakeethan** [comment]
@jawache out of the above list, many are listed on the tdp finder dataset with 0 tdp due to no data found about them mainly. 

Xeon Platinum 8176M is not found in the dataset entirely.

**Comment by jmcook1186** [comment]
Waiting on documentation before this can be closed out @gnanakeethan

**Comment by jmcook1186** [comment]
Closing as complete
